### ISSUE TITLE
fatal error C1083 

### ISSUE THREAD

**Issue by wuli12138** [esmini/esmini#78]
when i run 'cmake --build . --config Release --target install' command, the powershell will play no such file or directory, such as osi_common.pb.h osg/PositionAttitudeTransform gtest/gtest.h. i just download esmin source, then follow mkdir build  cd build..  but 4th command error, so what error is there? i am lack of dependency? thanks!

**Comment by esmini** [comment]
Thanks for reporting the issue. 
I assume you're on Windows, using PowerShell as command line tool. And I also assume you have a compiler installed, for example Visual Studio (with C++ SDK). 
Please perform the following steps exactly, and then provide the exact error message, and I will have a closer look.
1. Clone esmini (https://github.com/esmini/esmini.git) or download [the source code](https://github.com/esmini/esmini/archive/master.zip).
2. Open PowerShell and in root folder (where you have for example the README.md file)
3. `mkdir build`
4. `cd  build`
5. `cmake ..`
6. `cmake --build . --config Release --target install`

The dependencies (e.g. osi_common) should be downloaded in step 5, where the build scripts/solutions are created for your compiler environment.

**Comment by wuli12138** [comment]
yes, i run this project in win10 and have a compiler(vs 2019)
the command(cmake .. ) works well, but the folder(esimini/scripts) is the same as before running the command(cmake ..)

![微信截图_20210223083612](https://user-images.githubusercontent.com/41660074/108788136-44925000-75b2-11eb-9ed6-3054b0453c07.png)
 
there are some screenshots when i run command in powershell 
![微信截图_20210223082954](https://user-images.githubusercontent.com/41660074/108787901-b8802880-75b1-11eb-8210-106dca4889ce.png)
![微信截图_20210223083117](https://user-images.githubusercontent.com/41660074/108787907-b918bf00-75b1-11eb-8fff-4b37d32b389a.png)
![微信截图_20210223083146](https://user-images.githubusercontent.com/41660074/108787910-ba49ec00-75b1-11eb-84de-3eae09b73d3f.png)
![微信截图_20210223083201](https://user-images.githubusercontent.com/41660074/108787911-bae28280-75b1-11eb-9237-9a9f86b2aef9.png)
![微信截图_20210223083224](https://user-images.githubusercontent.com/41660074/108787916-bc13af80-75b1-11eb-9149-3dacf5ccb5f3.png)
![微信截图_20210223083235](https://user-images.githubusercontent.com/41660074/108787918-bcac4600-75b1-11eb-8d63-1093c2643009.png)
![微信图片_20210223083049](https://user-images.githubusercontent.com/41660074/108787920-bddd7300-75b1-11eb-8bc7-160951d70e23.png)
![微信截图_20210223083612](https://user-images.githubusercontent.com/41660074/108788124-3e9c6f00-75b2-11eb-876e-af10e57974e2.png)

thanks!

**Comment by wuli12138** [comment]
> Thanks for reporting the issue.
> I assume you're on Windows, using PowerShell as command line tool. And I also assume you have a compiler installed, for example Visual Studio (with C++ SDK).
> Please perform the following steps exactly, and then provide the exact error message, and I will have a closer look.
> 
> 1. Clone esmini (https://github.com/esmini/esmini.git) or download [the source code](https://github.com/esmini/esmini/archive/master.zip).
> 2. Open PowerShell and in esmini root folder (where you have for example the README.md file)
> 3. `mkdir build`
> 4. `cd  build`
> 5. `cmake ..`
> 6. `cmake --build . --config Release --target install`
> 
> The dependencies (e.g. osi_common) should be downloaded in step 5, where the build scripts/solutions are created for your compiler environment.



> Thanks for reporting the issue.
> I assume you're on Windows, using PowerShell as command line tool. And I also assume you have a compiler installed, for example Visual Studio (with C++ SDK).
> Please perform the following steps exactly, and then provide the exact error message, and I will have a closer look.
> 
> 1. Clone esmini (https://github.com/esmini/esmini.git) or download [the source code](https://github.com/esmini/esmini/archive/master.zip).
> 2. Open PowerShell and in esmini root folder (where you have for example the README.md file)
> 3. `mkdir build`
> 4. `cd  build`
> 5. `cmake ..`
> 6. `cmake --build . --config Release --target install`
> 
> The dependencies (e.g. osi_common) should be downloaded in step 5, where the build scripts/solutions are created for your compiler environment.

yes, i run this project in win10 and have a compiler(vs 2019)
the command(cmake .. ) works well, but the folder(esimini/scripts) is the same as before running the command(cmake ..)

![微信截图_20210223083612](https://user-images.githubusercontent.com/41660074/108788136-44925000-75b2-11eb-9ed6-3054b0453c07.png)
 
there are some screenshots when i run command in powershell 
![微信截图_20210223082954](https://user-images.githubusercontent.com/41660074/108787901-b8802880-75b1-11eb-8210-106dca4889ce.png)
![微信截图_20210223083117](https://user-images.githubusercontent.com/41660074/108787907-b918bf00-75b1-11eb-8fff-4b37d32b389a.png)
![微信截图_20210223083146](https://user-images.githubusercontent.com/41660074/108787910-ba49ec00-75b1-11eb-84de-3eae09b73d3f.png)
![微信截图_20210223083201](https://user-images.githubusercontent.com/41660074/108787911-bae28280-75b1-11eb-9237-9a9f86b2aef9.png)
![微信截图_20210223083224](https://user-images.githubusercontent.com/41660074/108787916-bc13af80-75b1-11eb-9149-3dacf5ccb5f3.png)
![微信截图_20210223083235](https://user-images.githubusercontent.com/41660074/108787918-bcac4600-75b1-11eb-8d63-1093c2643009.png)
![微信图片_20210223083049](https://user-images.githubusercontent.com/41660074/108787920-bddd7300-75b1-11eb-8bc7-160951d70e23.png)
![微信截图_20210223083612](https://user-images.githubusercontent.com/41660074/108788124-3e9c6f00-75b2-11eb-876e-af10e57974e2.png)

thanks!

**Comment by eknabevcc** [comment]
This was a real challenge!
First, the "esmini/scripts" folder is not expected to be modified in any way. The binaries should end up in the "esmini/bin" folder.

Now, the issue. One observation I make is that, for some strange reason, there is an unexpected space character in the file path for the "missing" header files. For example: " osi_common.pb.h". 
I have no clue where it comes from. Trying to reproduce: If I remove or rename that file I get the following similar error message:
```
C:\tmp\esmini (test)\esmini  (2)\EnvironmentSimulator\Modules\ScenarioEngine\SourceFiles\OSIReporter.cpp(15,10): fatal error
 C1083: Cannot open include file: 'osi_common.pb.h': No such file or directory [C:\tmp\esmini (test)\esmini  (2)\build\Envir
onmentSimulator\Modules\ScenarioEngine\ScenarioEngine.vcxproj]
```
As you can see there is no added space character in my error message. 
Further, I see that you have space in the folder path. Initially I thought that might be a root cause, but I added some spaces as well, and it seems to be no issue with that., which is good - it should be OK to have space in folder names.

So, in summary: For some reason the path to the header files of external dependencies gets an additional, unwanted leading " " (space) character. I haven't found the root cause to this behavior. But I will continue search for it. I still wanted to update you - for your information and perhaps you get some idea of root case?

**Comment by wuli12138** [comment]
thanks, i know you means. yesterday i unzip esmini folder, beacuse there is a esmini folder already, so when i unzip, it rename to esmini(2).
now, i unzip to a root path, when i run those command, it also report above error 

`E:\esmini\EnvironmentSimulator\Modules\ScenarioEngine\SourceFiles\OSIReporter.cpp(15,10): fatal error C1083: 无法打开包 括文件:
“osi_common.pb.h”: No such file or directory [E:\esmini\build\EnvironmentSimulator\Modules\ScenarioEngine\ScenarioEngin
e.vcxproj]`

when i run "cmake ..", it will download osg.7z osi.7z sumo.7z googletest.7z model.7z. the header files are in those files? because i can not find those unzip files, it lead to report this error c1083? 

and i run this project in liunx shell(windows10 wsl), it also reports similar error
`/home/wuli12138/esmini/EnvironmentSimulator/Modules/Controllers/ControllerSumo.cpp:19:10: fatal error: utils/geom/PositionVector.h: No such file or directory
   19 | #include <utils/geom/PositionVector.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.`

so i thinks my project still miss header files.
thanks!

**Comment by eknabevcc** [comment]
That was one interesting and good step forward. Looking carefully I now see that the leading " " (space) is gone from the header file path. `...fatal error C1083: 无法打开包 括文件: “osi_common.pb.h”: N..`. instead of the former `... “ osi_common.pb.h"...`. 

How it depends on the esmini root folder name I don't understand. But let's move forward. 

You're correct, the 7z (zip) files includes all headers and libraries of dependent software. They should end up in the folder `esmini/externals` with the following sub folders:
```
googletest
OpenSceneGraph
OSI
pugixml
SUMO
```

Can you check if you have anything in there? 

I'm not familiar with wsl, but I suspect that the cmake script should identify the environment as Linux and hence download Linux versions of the headers and libs. So make sure you start from scratch with an empty build folder before running the `cmake ..` command. By the way, also remove the external folder (and all subfolders) if it exists already, otherwise esmini thinks the correct files are  already downloaded. On Linux, next level subfolders should be named linux, e.g: OSI/linux while on Windows/Visual Studio they are named v10 (as version of SDK), e.g: OSI/v10.

But moving back to the native Windows 10 and Visual Studio, could you verify that the 7z files are not only downloaded but also extracted. Check log in the command window which should look something like (note the "-- extracting" lines):

```
PS C:\tmp\esmini (test)\esmini\build> cmake ..
-- Building for: Visual Studio 16 2019
-- CMake version: 3.19.0-rc3
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.18363.
-- The C compiler identification is MSVC 19.28.29337.0
-- The CXX compiler identification is MSVC 19.28.29337.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Configuring for Windows SDK v10 x64
-- downloading osg.7z ...
-- extracting osg.7z ...
-- downloading osi.7z ...
-- extracting osi.7z ...
-- downloading sumo.7z ...
-- extracting sumo.7z ...
-- downloading googletest.7z ...
-- extracting googletest.7z ...
-- downloading models.7z ...
-- extracting models.7z ...
Submodule 'test/OSC-ALKS-scenarios' (https://github.com/arauschert/OSC-ALKS-scenarios.git) registered for path '../test/OSC-ALKS-scenarios'
Cloning into 'C:/tmp/esmini (test)/esmini/test/OSC-ALKS-scenarios'...
Submodule path '../test/OSC-ALKS-scenarios': checked out '0ce2c9e23131dc90266fc0298060143f06040b88'
-- Configuring done
-- Generating done
-- Build files have been written to: C:/tmp/esmini (test)/esmini/build
```

If not, remove any existing "externals" folder and try "cmake .." again.

Finally, old versions of cmake does not support 7z . So could you also check what version of cmake you have? From shell, run `cmake --version`. I get:
```
PS C:\tmp\esmini (test)\esmini\build> cmake --version
cmake version 3.19.0-rc3

CMake suite maintained and supported by Kitware (kitware.com/cmake).
PS C:\tmp\esmini (test)\esmini\build>
```
I hope you can check these steps, and report back. I would like to understand and have the issue solved. Thanks.

**Comment by wuli12138** [comment]
> That was one interesting and good step forward. Looking carefully I now see that the leading " " (space) is gone from the header file path. `...fatal error C1083: 无法打开包 括文件: “osi_common.pb.h”: N..`. instead of the former `... “ osi_common.pb.h"...`.
> 
> How it depends on the esmini root folder name I don't understand. But let's move forward.
> 
> You're correct, the 7z (zip) files includes all headers and libraries of dependent software. They should end up in the folder `esmini/externals` with the following sub folders:
> 
> ```
> googletest
> OpenSceneGraph
> OSI
> pugixml
> SUMO
> ```
> 
> Can you check if you have anything in there?
> 
> I'm not familiar with wsl, but I suspect that the cmake script should identify the environment as Linux and hence download Linux versions of the headers and libs. So make sure you start from scratch with an empty build folder before running the `cmake ..` command. By the way, also remove the external folder (and all subfolders) if it exists already, otherwise esmini thinks the correct files are already downloaded. On Linux, next level subfolders should be named linux, e.g: OSI/linux while on Windows/Visual Studio they are named v10 (as version of SDK), e.g: OSI/v10.
> 
> But moving back to the native Windows 10 and Visual Studio, could you verify that the 7z files are not only downloaded but also extracted. Check log in the command window which should look something like (note the "-- extracting" lines):
> 
> ```
> PS C:\tmp\esmini (test)\esmini\build> cmake ..
> -- Building for: Visual Studio 16 2019
> -- CMake version: 3.19.0-rc3
> -- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.18363.
> -- The C compiler identification is MSVC 19.28.29337.0
> -- The CXX compiler identification is MSVC 19.28.29337.0
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> Configuring for Windows SDK v10 x64
> -- downloading osg.7z ...
> -- extracting osg.7z ...
> -- downloading osi.7z ...
> -- extracting osi.7z ...
> -- downloading sumo.7z ...
> -- extracting sumo.7z ...
> -- downloading googletest.7z ...
> -- extracting googletest.7z ...
> -- downloading models.7z ...
> -- extracting models.7z ...
> Submodule 'test/OSC-ALKS-scenarios' (https://github.com/arauschert/OSC-ALKS-scenarios.git) registered for path '../test/OSC-ALKS-scenarios'
> Cloning into 'C:/tmp/esmini (test)/esmini/test/OSC-ALKS-scenarios'...
> Submodule path '../test/OSC-ALKS-scenarios': checked out '0ce2c9e23131dc90266fc0298060143f06040b88'
> -- Configuring done
> -- Generating done
> -- Build files have been written to: C:/tmp/esmini (test)/esmini/build
> ```
> 
> If not, remove any existing "externals" folder and try "cmake .." again.
> 
> Finally, old versions of cmake does not support 7z . So could you also check what version of cmake you have? From shell, run `cmake --version`. I get:
> 
> ```
> PS C:\tmp\esmini (test)\esmini\build> cmake --version
> cmake version 3.19.0-rc3
> 
> CMake suite maintained and supported by Kitware (kitware.com/cmake).
> PS C:\tmp\esmini (test)\esmini\build>
> ```
> 
> I hope you can check these steps, and report back. I would like to understand and have the issue solved. Thanks.

thanks very much. now i have solved it. the main reason is my network. i can not download osg.7z osi.7z etc. so i download *.7z manually, then copy *.7z to /externals and modify the cmakelist.txt just to extract *.7z. now i have run it successfully . thanks!

**Comment by eknabevcc** [comment]
Wow, that was a good finding. Maybe the cmake scripts should warn in some way. But anyway, really good to understand the issue and good news that you solved it 👍 
I close the ticket for now.
